modules/packages/DistributedBag.chpl

```diff
@@ -64,631 +62,791 @@ use BlockDist;
+            bufferOffset += nTransfer;
+          }
+        }
+
+        // With the excess elements, redistribute it...
+        bufferOffset = 0;
+        for loc in localThis.targetLocales do on loc {
+          var average = avg;
+          ref segment = getPrivatizedThis.bag.segments[segmentIdx];
+          var nElems = segment.nElems.read() : int;
+          if average > nElems {
+            var give = average - nElems;
+            var arr : [1..give] eltType;
+            on bufferOffset {
+              var tmpBuffer = buffer;
+              for i in 1 .. give {
+                arr[i] = tmpBuffer[bufferOffset];
+                bufferOffset += 1;
+              }
+            }
             for i in 1 .. give {
-              arr[i] = tmpBuffer[bufferOffset];
-              bufferOffset += 1;
+              segment.addElements(arr[i]);
             }
           }
-          for i in 1 .. give {
-            segment.addElements(arr[i]);
-          }
         }
-      }
 
-      // Lastly, if there are items left over, just add them to our locale's segment.
-      if excess > bufferOffset {
-        ref segment = localThis.bag.segments[segmentIdx];
-        var give = excess - bufferOffset;
-        var tmpBuffer = buffer + bufferOffset;
-        segment.addElementsPtr(tmpBuffer, give, buffer.locale.id);
+        // Lastly, if there are items left over, just add them to our locale's segment.
+        if excess > bufferOffset {
+          ref segment = localThis.bag.segments[segmentIdx];
+          var nLeftOvers = excess - bufferOffset;
+          var tmpBuffer = buffer + bufferOffset;
+          segment.addElementsPtr(tmpBuffer, nLeftOvers, buffer.locale.id);
+        }
+
+        c_free(buffer);
       }
 
-      c_free(buffer);
+      // Phase 3: Release all locks from first node and segment to last node and segment.
+      for loc in localThis.targetLocales do on loc {
+        var instance = getPrivatizedThis;
+        for segmentIdx in 0 .. #here.maxTaskPar {
+          ref segment = instance.bag.segments[segmentIdx];
+          segment.releaseStatus();
+        }
+      }
     }
 
-    // Phase 3: Release all locks from first node and segment to last node and segment.
-    for loc in localThis.targetLocales do on loc {
-      var instance = getPrivatizedThis;
-      for segmentIdx in 0 .. #here.maxTaskPar {
-        ref segment = instance.bag.segments[segmentIdx];
-        segment.releaseStatus();
+    /*
+      Iterate over each bag in each node. To avoid holding onto locks, we take
+      a snapshot approach, increasing memory consumption but also increasing parallelism.
+      This allows other concurrent, even mutating, operations while iterating,
+      but opens the possibility to iterating over duplicates or missing elements
+      from concurrent operations.
+
+      .. note:: 
+
+        `zip` iteration is not yet supported with rectangular data structures.

```

Comment:
Let's rewrite this. Something like "This data structure cannot be used in a zippered iteration." Just to make it more general.

Reply:
Can I potentially defer this until next commit? I have a ton of work to do now that school started, I'll promise to come back to it later (as comment said its incremental revision)

Reply:
This one is probably the simplest one :) But anyways, I think it would be nice if it is addressed before the release at least (sometime in the Fall, I guess)

Reply:
Speaking of, when is the next release? You're right in that I need to have my data structures improved by then, but I'd like to know when.

Reply:
Feature freeze is Sept 18th, doc/code freeze is Sept 25th.

Reply:
@mppf 

This month?! So soon after GSoC too... If you were asking whether or not this should be in official release, I'd say 'no'. I can't commit enough time to have it completely bug-free, not to mention there are other gating issues as well which would require my code to change once they are resolved. Everything about it hits places which are bug-ridden, both in terms of the implementation and due to Chapel itself, and I can imagine users running into many more that I haven't in the limited coverage of my tests.

With that said, the data structures make amazing additions to the language, but if anything I'd label them 'experimental' right now, and should remain in modules/packages.